File: CMFTAspNet/Models/Project.cs
using CMFTAspNet.Services.Interfaces;
using System.ComponentModel.DataAnnotations.Schema;

namespace CMFTAspNet.Models
{
    public class Project : WorkTrackingSystemOptionsOwner<Project>, IEntity
    {
        public int Id { get; set; }

        public string Name { get; set; }
        
        public SearchBy SearchBy { get; set; }
        
        public List<string> WorkItemTypes { get; set; } = new List<string> { "Epic" };

        [NotMapped]
        public IEnumerable<Team> InvolvedTeams => Features.SelectMany(f => f.RemainingWork).Select(rw => rw.Team).Distinct();

        public string SearchTerm { get; set; }

        public List<Feature> Features { get; } = new List<Feature>();

        public List<Milestone> Milestones { get; } = new List<Milestone>();

        public bool IncludeUnparentedItems { get; set; }

        public int DefaultAmountOfWorkItemsPerFeature { get; set; } = 25;

        public DateTime ProjectUpdateTime { get; set; }

        public void UpdateFeatures(IEnumerable<Feature> features)
        {
            Features.Clear();
            Features.AddRange(features);

            ProjectUpdateTime = DateTime.Now;
        }
    }

    public enum SearchBy
    {
        Tag,
        AreaPath,
    }
}

File: Lighthouse.Frontend/src/models/Project.ts
import { Feature } from './Feature';
import { IFeatureOwner } from './IFeatureOwner';
import { IMilestone } from './Milestone';
import { ITeam, Team } from './Team'

export interface IProject extends IFeatureOwner {
    name: string;
    id: number;
    involvedTeams: ITeam[];
    lastUpdated: Date;
    milestones: IMilestone[];
}

export class Project implements IProject {
    name!: string;
    id!: number;

    features!: Feature[];
    involvedTeams!: Team[];
    milestones!: IMilestone[];

    lastUpdated!: Date;

    constructor(name: string, id: number, involvedTeams: Team[], features: Feature[], milestones: IMilestone[], lastUpdated: Date) {
        this.name = name;
        this.id = id;
        this.involvedTeams = involvedTeams;
        this.lastUpdated = lastUpdated;
        this.features = features || [];
        this.milestones = milestones;
    }

    get remainingWork(): number {
        const totalRemainingWork = this.features.reduce((acc, feature) => acc + feature.getAllRemainingWork(), 0);
    
        return totalRemainingWork;
    }

    get remainingFeatures(): number {
        return this.features.length;
    }
}